cava/CoqStyleGuide.md

```diff
@@ -0,0 +1,254 @@
+
+### Line length
+
+- Maximum line length 80 characters.
+  * Many Coq IDE setups divide the screen in half vertically and use only half
+    to display source code, so more than 80 characters can be genuinely hard to
+    read on a laptop.
+
+### Whitespace and indentation
+
+- No trailing whitespace.
+- Spaces, not tabs.
+- Files should end with a newline.
+  * Many editors do this automatically on save.
+- Default indentation is 2 spaces.
+  * Keeping this small prevents complex proofs from being indented ridiculously
+    far, and matches IDE defaults.
+- Use 2-space indents if inserting a line break immediately after:
+  * `Proof.`
+  * `fun <...> =>`
+  * `forall <...>,`
+  * `exists <....>,`
+- The style for indenting arguments in function application depends on where
+  you make a line break.  If you make the line break immediately after the
+function name, use a 2-space indent.  However, if you make it after one or more
+arguments, align the next line with the first argument:
+  ```coq
+  (Z.pow
+     1 2)
+  (Z.pow 1 2 3
+         4 5 6)
+  ```
+- `Inductive` cases should not be indented. Example:
+  ```coq
+  Inductive Foo : Type :=
+  | FooA : Foo
+  | FooB : Foo
+  .
+  ```
+- `match` or `lazymatch` cases should line up with the "m" in `match` or "l" in
+  `lazymatch`, as in the following examples:
+  ```coq
+  match x with
+  | 3 => true
+  | _ => false
+  end.
+
+  lazymatch x with
+  | 3 => idtac
+  | _ => fail "Not equal to 3:" x
+  end.
+
+  repeat match goal with
+         | _ => progress subst
+         | _ => reflexivity
+         end.
+
+  do 2 lazymatch goal with
+       | |- context [eq] => idtac
+       end.
+  ```
+
+## Definitions and Fixpoints
+
+- It's okay to leave the return type of `Definition`s and `Fixpoint`s implicit
+  (e.g. `Definition x := 5` instead of `Definition x : nat := 5`) when the type
+is very simple or obvious (for instance, the definition is in a file which deals
+exclusively with operations on `Z`).
+
+## Lemma/Theorem statements
+
+- Generally, use `Theorem` for the most important, top-level facts you prove
+  and `Lemma` for everything else.
+- Insert a line break after the colon in the lemma statement.
+- Insert a line break after the comma for `forall` or `exist` quantifiers.
+- Implication arrows (`->`) should share a line with the previous hypothesis,
+  not the following one.
+- There is no need to make a line break after every `->`; short preconditions
+  may share a line.
+
+## Proofs and tactics
+
+- Use the `Proof` command (lined up vertically with `Lemma` or `Theorem` it
+  corresponds to) to open a proof, and indent the first line after it 2 spaces.
+- Very small proofs (where `Proof. <tactics> Qed.` is <= 80 characters) can go
+  all in one line.
+- When ending a proof, align the ending statement (`Qed`, `Admitted`, etc.)
+  with `Proof`.
+- Avoid using autogenerated names (e.g. `H0`, `n0`). They make the proof
+  break when new hypotheses are added or variable names change.
```

Comment:
Does this mean all instances of e.g `intros` should include new identifiers- is it ok to omit when the proof is immediately dispatched e.g. `intros; my_solver`?

Reply:
Ah, I should clarify this a bit. To me, it doesn't matter whether Coq generates automatic names, it matters whether the proofs explicitly *refer* to them, because then the proofs will break if the automatic names change. So `intros; my_solver` is fine; `intros; apply H; my_solver` is not. Will reword this.

Reply:
I've changed the wording, mind taking another look at this bullet?